x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_count_distinct.csv-spec

```diff
@@ -177,3 +177,70 @@ m:long  | languages:i
 20   | 5
 10   | null
 ;
+
+foldablePrecision
+required_capability: extend_aggs_on_constants_support
+from employees | eval x = 82+8 | stats count_distinct(salary, x*100);
+
+count_distinct(salary, x*100):long
+100
+;
+
+foldablePrecisionWithGroupings
+required_capability: extend_aggs_on_constants_support
+from employees | eval x = 82+8 | stats cd1=count_distinct(salary, x*100), cd2=count_distinct(salary, 10+10) by gender, languages | sort gender, languages;
+
+   cd1:long    |    cd2:long   |gender:keyword |languages:integer   
+4              |4              |F              |1              
+5              |5              |F              |2              
+6              |6              |F              |3              
+6              |6              |F              |4              
+9              |9              |F              |5              
+3              |3              |F              |null           
+9              |9              |M              |1              
+11             |11             |M              |2              
+11             |11             |M              |3              
+11             |11             |M              |4              
+8              |8              |M              |5              
+7              |7              |M              |null           
+2              |2              |null           |1              
+3              |3              |null           |2              
+1              |1              |null           |4              
+4              |4              |null           |5              
+;
+
+foldableNullPrecision
+required_capability: extend_aggs_on_constants_support
+from employees | eval x = 82+null | stats count_distinct(salary, x*100), count_distinct(salary, null), count_distinct(salary, null + 1);
+
+count_distinct(salary, x*100):long|count_distinct(salary, null):long|count_distinct(salary, null + 1):long
+100                               |100                              |100
```

Comment:
I am not so sure :-), now that you mentioned this aspect. Maybe we want null here, but at the same time `count_distinct` and `count` DO make sense to be not nullable, counting something should always be something concrete.

If `null` is not the default precision and we cannot return `null` as the return value of `count_distinct`, should we error out then? (like weighted_avg and percentile)

Reply:
count are _special_ in that they are not nullable - they always return a value. If anything, MV_COUNT should be aligned with it.
It's up to the function to decide if null is treated as "default value" or invalid. Since there's no COUNT_DISTINCT(field), I'd opt for null meaning default.

Reply:
I'd be in favor of `COUNT_DISTINCT(field, null)` resulting in an invalid query exception via the `Validatable` interface - but I'm fine with default precision, as long as we document this clearly :)